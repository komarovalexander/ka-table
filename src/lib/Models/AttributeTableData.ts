import { HTMLAttributes } from 'react';

import { ChildProps, DispatchFunc } from '../types';

export class AttributeTableData {
  public baseFunc!: any;
  public childElementAttributes!: HTMLAttributes<HTMLElement>;
  public childProps!: ChildProps;
  public dispatch!: DispatchFunc;
}
