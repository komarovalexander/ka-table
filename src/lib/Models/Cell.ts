/**
 * Describes the position of cell in table
 */
export class Cell {
  /*
  * Specifies the field of data
  */
  public columnKey!: string;

  /*
  * Specifies the value of row's key where cell is placed
  */
  public rowKey!: any;
}
