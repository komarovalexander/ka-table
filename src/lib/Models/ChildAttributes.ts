
import { ITableAllProps } from '../';
import { ICellContentProps } from '../Components/CellContent/CellContent';
import { ChildAttributesItem } from '../types';

export class ChildAttributes {
  public cell?: ChildAttributesItem<ICellContentProps>;
  public table?: ChildAttributesItem<ITableAllProps>;
}
