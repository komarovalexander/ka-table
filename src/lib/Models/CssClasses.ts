export class CssClasses {
  public table?: string = 'ka-table';
  public tbody?: string = 'ka-tbody';
  public thead?: string = 'ka-thead';
  public theadRow?: string = 'ka-tr ka-thead-row';
  public theadCell?: string = 'ka-thead-cell';
  public groupRow?: string = 'ka-tr ka-group-row';
  public row?: string = 'ka-tr ka-row';
  public rowSelected?: string = 'ka-row-selected';
  public cell?: string = 'ka-cell';

  public iconGroupArrowCollapsed?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-collapsed';
  public iconGroupArrowExpanded?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-expanded';
  public iconClose?: string = 'ka-icon ka-icon-close';
  public iconSortArrowDown?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-down';
  public iconSortArrowUp?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-up';

  public checkbox ?: string = 'ka-input';
  public dateInput ?: string = 'ka-input';
  public numberInput ?: string = 'ka-input';
  public textInput ?: string = 'ka-input';
}
