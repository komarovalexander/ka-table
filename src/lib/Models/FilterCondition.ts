/* Filter condition for specific field */
export class FilterCondition {
  public field!: string;
  public operator!: string;
  public value!: any;
}
