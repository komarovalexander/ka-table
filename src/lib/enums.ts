export enum DataType {
  Boolean = 'boolean',
  Date = 'date',
  Number = 'number',
  Object = 'object',
  String = 'string',
}

export enum EditingMode {
  None = 'none',
  Cell = 'cell',
}

export enum ActionType {
  CloseEditor = 'CloseEditor',
  ChangeFilterRow = 'ChangeFilterRow',
  OpenEditor = 'OpenEditor',
  ChangeRowData = 'ChangeRowData',
  DeselectRowData = 'DeselectRowData',
  SelectRow = 'SelectRow',
  ScrollTable = 'ScrollTable',
  ChangeSorting = 'ChangeSorting',
}

export enum KeyboardEnum {
  Esc = 27,
  Enter = 13,
}

export enum SortDirection {
  Ascend = 'ascend',
  Descend = 'descend',
}

export enum SortingMode {
  None = 'none',
  Single = 'single',
}

export enum FilteringMode {
  None = 'none',
  FilterRow = 'filterRow',
}

export enum FilterOperatorName {
  Equal = '=',
  MoreThan = '>',
  LessThan = '<',
  MoreThanOrEqual = '>=',
  LessThanOrEqual = '<=',
  Contains = 'contains',
}
