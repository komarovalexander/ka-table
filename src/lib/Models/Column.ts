import { DataType, SortDirection } from '../enums';
import {
  CellFunc, EditorFunc, FilterRowFunc, FormatFunc, HeaderCellFunc, SearchFunc, ValidationFunc,
} from '../types';

/**
 * Describes column of table its look and behaviour
 */
export class Column {
  public cell?: CellFunc;
  public dataType?: DataType;
  public editor?: EditorFunc;
  public filterRowCell?: FilterRowFunc;
  public filterRowValue?: any;
  public filterRowOperator?: any;
  public field?: string;
  public format?: FormatFunc;
  public headCell?: HeaderCellFunc;
  public isEditable?: boolean;
  public key!: string;
  public search?: SearchFunc;
  public sortDirection?: SortDirection;
  public style?: React.CSSProperties;
  public title?: string;
  public validation?: ValidationFunc;
}
