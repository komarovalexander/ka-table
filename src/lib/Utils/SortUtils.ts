import { SortDirection } from '../enums';
import { Column } from '../Models/Column';

export const sortData = (columns: Column[], data: any): any[] => {
  const sortedColumn = columns.find((column) => column.sortDirection);
  if (!sortedColumn) { return data; }
  const columnKey = sortedColumn.key;
  const sortFunc = sortedColumn.sortDirection === SortDirection.Ascend ?
    ((a: any, b: any) => a[columnKey] < b[columnKey] ? -1 : 1)
    : ((a: any, b: any) => a[columnKey] > b[columnKey] ? -1 : 1);
  const newData = [...data].sort(sortFunc);
  return newData;
};
