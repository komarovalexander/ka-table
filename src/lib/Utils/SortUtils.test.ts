import { DataType, SortDirection } from '../enums';
import { Column } from '../Models/Column';
import { sortData } from './SortUtils';

const data: any[] = [
  { column: 1, id: 1 },
  { column: 3, id: 2 },
  { column: 2, id: 3 },
  { column: 3, id: 4 },
  { column: 3, id: 5 },
];

const columns: Column[] = [
  { key: 'id', title: 'Id', dataType: DataType.String },
  { key: 'column', title: 'Column 1', dataType: DataType.String, sortDirection: SortDirection.Descend },
];

describe('sortData', () => {
  it('should not change original data', () => {
    const newData = sortData(columns, data);
    expect(newData).not.toBe(data);
  });

  it('should be sorted by Descend', () => {
    const newData = sortData(columns, data);
    expect(newData).toMatchSnapshot();
  });

  it('should be sorted by Ascend', () => {
    const columns2 = [
      { key: 'column', title: 'Column 1', sortDirection: SortDirection.Ascend, dataType: DataType.String },
    ];
    const newData = sortData(columns2, data);
    expect(newData).toMatchSnapshot();
  });
});
