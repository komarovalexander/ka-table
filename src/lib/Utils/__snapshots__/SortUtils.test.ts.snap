// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortData should be sorted by Ascend 1`] = `
Array [
  Object {
    "column": 1,
    "id": 1,
  },
  Object {
    "column": 2,
    "id": 3,
  },
  Object {
    "column": 3,
    "id": 2,
  },
  Object {
    "column": 3,
    "id": 4,
  },
  Object {
    "column": 3,
    "id": 5,
  },
]
`;

exports[`sortData should be sorted by Descend 1`] = `
Array [
  Object {
    "column": 3,
    "id": 2,
  },
  Object {
    "column": 3,
    "id": 4,
  },
  Object {
    "column": 3,
    "id": 5,
  },
  Object {
    "column": 2,
    "id": 3,
  },
  Object {
    "column": 1,
    "id": 1,
  },
]
`;
